### ISSUE TITLE
Feature Request - Retention by Available Disk Space

### ISSUE THREAD

**Issue by XtremeOwnageDotCom** [blakeblackshear/frigate#994]
Reading through the documentation, and the available configuration parameters-

I do not see the ability to say, allocate 3TB to recordings. 

In a typical NVR setup, you would typically dedicate a single, large disk  to your NVR system. In my current setup, I use a 8TB disk in conjunction with Blue Iris.

I can set specific retention times on a per camera/per folder basis if required, but, for my recordings, I would prefer to maximize my retention based on the amount of available disk space I have. When the disk starts filling up, delete the oldest recordings to free up space.

TLDR;

Asking for a new feature to control retention by disk space, rather then only by a static number of days.

**Comment by rhatguy** [comment]
+1 for this.  Many use cases would want to just keep as much as possible in a given amount of space.

**Comment by XtremeOwnageDotCom** [comment]
+ Further Activity. Still a REALLY good.... idea to add.

**Comment by OptimusGREEN** [comment]
yeah this would be super handy. My feeds have actually stopped showing in home assistant and after some debugging it appears to me my nvr drive is full.

`
[2021-11-07 18:59:33] ffmpeg.driveway.record_rtmp    ERROR   : Could not write header for output file #0 (incorrect codec parameters ?): No space left on device`

**Comment by Bod1985** [comment]
This would be really useful. I use a docker inside a Debian VM with a second hard drive passed through. 

This drive became full today causing IO errors and crashed my entire Debian VM. (I can make improvements elsewhere to prevent this in future but it also highlighted my desire for this feature in Frigate)

**Comment by mnaz** [comment]
I would agree it would be nice to have a setting to help protect the amount stored to avoid filling the drive accidentally.

**Comment by bagobones** [comment]
Doing this with the option for fixed space used, % of available disk, or minimum free space to keep would be ideal.. Not all NVRs implement all 3 options but most have some form of roll over based on space. These options are particularly used full if you are storing on a shared NAS where fixed max might be best, or if you just want to target a disk and leave a little space free.

**Comment by user897943** [comment]
This is definitely needed. I don't want the smart AI stuff, that should be optional. I installed Frigate because it integrates with HA so well and I like the interface. I thought it was very reliable up to today when it crashed with a IO error because the external HDD I use for the recordings was full.

**Comment by Vitani** [comment]
I'm using this Powershell script, run on a 15 minute schedule, to tidy up my recordings, hope it helps someone else (just change the `$drive` and `$target` variables). Obviously I do not provide any warranty, and use at your own risk.

```
$drive = "D:"
$target = "$drive\Frigate\Media\recordings"
$desiredGB = 100 * 1GB
$freeSpace = ([wmi]"Win32_LogicalDisk.DeviceID='$drive'").FreeSpace
if ($freeSpace -ge $desiredGB) {
	Write-Host "Free space is $([math]::Round($freeSpace / 1GB))GB/$([math]::Round($desiredGB / 1GB))GB. Nothing to do."
	exit
}

$spaceNeeded = $desiredGB - $freeSpace
Write-Host "Need to free up $([math]::Round($spaceNeeded / 1GB))GB, Enumerating files ..."

$files = Get-ChildItem $target -File -Recurse | Sort-Object CreationTime
$spaceSaved = 0mb
$filesToDelete = foreach ($file in $files) {
	$file #return the file so it gets added to $filesToDelete
	$spaceSaved += $file.length
	if ($spaceSaved -ge $spaceNeeded) {
		break
	}
}
if ($filesToDelete.Count -eq 0) {
	Write-Warning "Done; No files available to delete. Further action required."	
}
else {
	Write-Host "Done. Deleting $($filesToDelete.Count) files ..."
	$filesToDelete | Remove-Item -Force -Confirm:$false -Verbose
	
	$tailRecursion = {
		param(
			$Path
		)
		foreach ($childDirectory in Get-ChildItem -Force -LiteralPath $Path -Directory) {
			& $tailRecursion -Path $childDirectory.FullName
		}
		$currentChildren = Get-ChildItem -Force -LiteralPath $Path
		$isEmpty = $currentChildren -eq $null
		if ($isEmpty) {
			Write-Verbose "Removing empty folder at path '${Path}'." -Verbose
			Remove-Item -Force -LiteralPath $Path
		}
	}	
	
	Write-Host "Done, deleting empty folders ..."
	& $tailRecursion -Path $target
	
	Write-Host "Done, waiting 30 seconds before checking free space again ..."
	Start-Sleep -Seconds 30
	$freeSpace = ([wmi]"Win32_LogicalDisk.DeviceID='$drive'").FreeSpace
	if ($freeSpace -lt $desiredGB) {
		Write-Warning ("Files deleted only freed {0}GB of {1}GB needed. Further action required." -f [math]::Round($spaceSaved / 1GB), [math]::Round($spaceNeeded / 1GB))
	}
	else {
		Write-Host "All done, $([math]::Round($freeSpace / 1GB))GB of free space available."
	}
}
```

**Comment by einarhauks** [comment]
Please either implement this or allow us to configure some maximum space for Frigate recordings. When running Frigate as an Add-On in Home Assistant OS, one cannot simply assign recordings to another volume (please correct me if I'm wrong). Having Frigate filling up the Hassio disk is a real pain...

**Comment by NPutting** [comment]
> I'm using this Powershell script, run on a 15 minute schedule, to tidy up my recordings, hope it helps someone else (just change the `$drive`, `$target` and `$desiredGB` variables). Obviously I do not provide any warranty, and use at your own risk.
> 
> ```
> $drive = "D:"
> $target = "$drive\Frigate\Media\recordings"
> $desiredGB = 100 * 1GB
> 
> $freeSpace = ([wmi]"Win32_LogicalDisk.DeviceID='$drive'").FreeSpace
> if ($freeSpace -ge $desiredGB) {
> 	Write-Host "Free space is $([math]::Round($freeSpace / 1GB))GB/$([math]::Round($desiredGB / 1GB))GB. Nothing to do."
> 	exit
> }
> 
> $spaceNeeded = $desiredGB - $freeSpace
> Write-Host "Need to free up $([math]::Round($spaceNeeded / 1GB))GB, Enumerating files ..."
> 
> $files = Get-ChildItem $target -File -Recurse | Sort-Object CreationTime
> $spaceSaved = 0mb
> $filesToDelete = foreach ($file in $files) {
> 	$file #return the file so it gets added to $filesToDelete
> 	$spaceSaved += $file.length
> 	if ($spaceSaved -ge $spaceNeeded) {
> 		break
> 	}
> }
> if ($filesToDelete.Count -eq 0) {
> 	Write-Warning "Done; No files available to delete. Further action required."	
> }
> else {
> 	Write-Host "Done. Deleting $($filesToDelete.Count) files ..."
> 	$filesToDelete | Remove-Item -Force -Confirm:$false -Verbose
> 	
> 	$tailRecursion = {
> 		param(
> 			$Path
> 		)
> 		foreach ($childDirectory in Get-ChildItem -Force -LiteralPath $Path -Directory) {
> 			& $tailRecursion -Path $childDirectory.FullName
> 		}
> 		$currentChildren = Get-ChildItem -Force -LiteralPath $Path
> 		$isEmpty = $currentChildren -eq $null
> 		if ($isEmpty) {
> 			Write-Verbose "Removing empty folder at path '${Path}'." -Verbose
> 			Remove-Item -Force -LiteralPath $Path
> 		}
> 	}	
> 	
> 	Write-Host "Done, deleting empty folders ..."
> 	& $tailRecursion -Path $target
> 	
> 	Write-Host "Done, waiting 30 seconds before checking free space again ..."
> 	Start-Sleep -Seconds 30
> 	$freeSpace = ([wmi]"Win32_LogicalDisk.DeviceID='$drive'").FreeSpace
> 	if ($freeSpace -lt $desiredGB) {
> 		Write-Warning ("Files deleted only freed {0}GB of {1}GB needed. Further action required." -f [math]::Round($spaceSaved / 1GB), [math]::Round($spaceNeeded / 1GB))
> 	}
> 	else {
> 		Write-Host "All done, $([math]::Round($freeSpace / 1GB))GB of free space available."
> 	}
> }
> ```

Do you have any problems with the DB becoming bloated when manually deleting the recordings?

**Comment by Vitani** [comment]
> Do you have any problems with the DB becoming bloated when manually deleting the recordings?

I didn't notice, but I have since moved to _Blue Iris_ as I found it more reliable & had more feature that I wanted to use.

**Comment by tehniemer** [comment]
In my mind this is a very necessary feature for those who keep 24/7 recordings. It would be much easier to set a storage limit than try to guess how many days it'll take to fill it up. In this case it would also make adding or removing cameras painless.

**Comment by ozett** [comment]
as advised i wil add my wish to this FR here, that an infomation about remaining recording-time would be great.

that would be highly dynamic, based on recent savings, deletion or blocked-deletions 
(favorites clips become delete-protected).
but some AI could kick in to make this estimations about remainig time, based on actual disc space - and historical data.
see https://github.com/blakeblackshear/frigate/pull/3225#issuecomment-1133064092

**Comment by elektrinis** [comment]
Frigate crashed my whole smart home while I was away on vacation, due to disk being full.
I can't believe there is no disk-related configuration, like to use full disk and set the locanion, at minimum. "Retention" setting is really not as useful, as free space is the limiting factor in most cases.

**Comment by bagobones** [comment]
> Segment sizes are quite consistent / predictable so figuring out how many days = GB is fairly easy to do.

If this statement was true then the original feature request would be trivial. 

As others pointed out it is not due to all the other disk using features. 

Simply we need the recorder to realize space is about to run out and start purging / rolling over old content when a set amount of free space is not available.

**Comment by user897943** [comment]
> > "Retention" setting is really not as useful, as free space is the limiting factor in most cases.
> 
> Segment sizes are quite consistent / predictable so figuring out how many `days = GB` is fairly easy to do.
> 
> That being said, the flexibility and convenience is recognized which is why this FR is pinned and will be included in a future release (hopefully 0.12 but we'll see)

Ok don't take this personal, none of you developing, but please STOP telling everyone that it's easy. It's not easy. I had data loss that messed things up for me really badly in a drug case. Do you guys want Frigate to be a bona-fide recorder or a hobby toy? I **love** the application but stop building silly features and saying "hopefully on 0.12". It should be the absolute priority. To have a whole system crash, not just Frigate, is unacceptable. User fault? Maybe. But a CCTV recording app that isn't aware of space is alpha. I want to politely ask you guys developing this app to reconsider your priorities and make space-based management happen ASAP.

**Comment by elektrinis** [comment]
This ^

**Comment by NickM-27** [comment]
> Given I have multiple different manufacturers of Cameras (Reolink, Hikvision, Amcrest, etc....), the encoding/resolution/bitrate will not be consistent across cameras. Some are 5MP / H265, others are lower resolution using H264, for example.

To be clear, I wasn't saying it was a great solution but suggesting how the "days" option can be utliized until this is supported. The segment size for each camera will be consistent, so even if one is h265 and one is h264 and they are different, each individual camera will have predictable sizes. 

> If this statement was true then the original feature request would be trivial.
>
> As others pointed out it is not due to all the other disk using features.

It is true, see [this comment](https://github.com/blakeblackshear/frigate/issues/2790#issuecomment-1036173831) for how to do it with your own camera.

Also see [this comment](https://github.com/blakeblackshear/frigate/issues/2790#issuecomment-1036156266) for a partial explanation for why this hasn't been done "ASAP".

It is not trivial to implement due to a large number of edge cases and considerations for setups with many cameras. 

> I love the application but stop building silly features and saying "hopefully on 0.12".

You noted in your previous issues that you do not use Frigates AI features / object detection at all. So calling the features that have been worked on "silly" because they do not apply to your particular use case is disingenuous. 

Frigate has many users with many different use cases / priorities, we have to do our best to cater to everyone as each person believes their feature is "top priority". I said "hopefully" as I am not in a position to promise anything for the next release cycle, but it is definitely a high priority feature that is recognized.

**Comment by esmoyer** [comment]
People do realize that this is not the only NVR software available, right? Frigate is free and developed by Blake and others giving their time to provide us with an NVR option with some pretty awesome features. I can't believe how far it has come.

I get putting in a feature request or idea, but, unless I'm interpreting things wrong, reading through some of these comments (leaning more towards demands), it sounds like some of you need to be shelling out $$$ for a commercial or enterprise system instead. Frigate just might not be for those of you who need or expect it to have every little x, y, z feature.

As some of these "requests" start turning into "demands", I sure hope y'all are making good use of that Donate/Sponsor link. Or, better yet, if it's so simple, fork the code, implement the feature yourself and submit a pull request to be merged.

**Comment by NickM-27** [comment]
closing as the original request for frigate cleaning up recordings when storage is full has been implemented since 0.12, subsequent more specific feature requests can be made.
